### PR TITLE
Move implementations out of header file

### PR DESCRIPTION
**Pull request by bfabio**
Move all the code to the cpp file where you expect it to be, so the actual implementation is not all in an header file.

### CODE REVIEW

**Comment by parsley** [comment]
Hi,

**tl;dr:**
is there a reason you keep the implementation within the header file?

**long version:**
I'd like to contribute changes I did to your project (especially a major change that fixes the checksum issue once and for all). However I've merged this PR in my version and since then I use the .cpp-file along with the .h-file. It's the way C/C++ should be, doesn't it?
Is there a real advantage in keeping the implementation within the header file?

If you are willing to go the ".h + .cpp way" I'm happy to prepare a PR that makes the switch for you and adds my contributions.
However if you have a good reason to keep it ".h only" I'll have to make up my mind to either convert my version back to the ".h only" way to be able to contribute or to further diverge from your project.

Obviously I'd prefer to agree on one way and I'm happy to go either with or without the .cpp but before I add work to convert back I want to make sure not to trash anything that you would rather like to keep. ;-)
